src/Momento.Sdk/Config/Middleware/FairAsyncSemaphore.cs

```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Momento.Sdk.Config.Middleware
+{
+    public class FairAsyncSemaphore
+    {
+        private readonly Channel<bool> _ticketChannel;
+
+        public FairAsyncSemaphore(int numTickets)
+        {
+            //Console.WriteLine($"Creating semaphore with {numTickets} tickets");
+            _ticketChannel = Channel.CreateBounded<bool>(numTickets);
+
+            for (var i = 0; i < numTickets; i++)
+            {
+                bool success = _ticketChannel.Writer.TryWrite(true);
+                if (!success)
+                {
+                    throw new ApplicationException("Unable to initialize async channel");
+                }
+            }
+        }
+
+        public async Task WaitOne()
+        {
+            //Console.WriteLine($"Waiting for semaphore");
+            await _ticketChannel.Reader.ReadAsync();
+        }
+
+        public async Task Release()
+        {
+            //Console.WriteLine($"Releasing for semaphore");
+            await _ticketChannel.Writer.WriteAsync(true);
```

Comment:
```suggestion
        public void Release()
        {
            var balanced = _ticketChannel.Writer.TryWrite(true);
            if (!balanced)
            {
                throw new ApplicationException("more releases than waits! These must be 1:1")
            }
```

if a user were to put in imbalanced releases then their coroutines will all stack up in memory and it will eventually OOM. If you do this instead then it still won't necessarily show the offending stack but it'll at least tell you promptly that something is wrong.

Also if you do it this way there's no task and state machine so it'll be a handful of µs faster

Reply:
good call, ty